### PR TITLE
Add GraphMultisetTransformer (Baek et al., ICLR 2021)

### PR DESCRIPTION
**Pull request by JinheonBaek**
This PR adds the GMT global pooling layer for graph representation learning, from

Accurate Learning of Graph Representations with Graph Multiset Pooling
Jinheon Baek*, Minki Kang*, and Sung Ju Hwang
https://arxiv.org/abs/2102.11533

Changes include the implementation of the GMT pooling layer, an example on the PROTEINS dataset, docs, and tests.

### CODE REVIEW

File: torch_geometric/nn/glob/gmt.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+            :obj:`['GMPool_G', 'GMPool_I']`,
+            :obj:`['GMPool_G', 'SelfAtt', 'GMPool_I']` or
+            :obj:`['GMPool_G', 'SelfAtt', 'SelfAtt', 'GMPool_I']`).
+            (default: :obj:`['GMPool_G', 'SelfAtt', 'GMPool_I']`)
+        num_heads (int, optional): Number of heads in attention.
+            (default: :obj:`4`)
+        ln (bool, optional): If set to :obj:`True`, will make use of
+            layer normalization. (default: :obj:`False`)
+    """
+    def __init__(self, in_channels, hidden_channels, out_channels,
```
Comment (rusty1s) [inline torch_geometric/nn/glob/gmt.py:126]: Can we add type hints? For example, `in_channels: int`, `hidden_channels: int`, ...

File: torch_geometric/nn/glob/gmt.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+                )
+                n_out_nodes = math.ceil(n_out_nodes * self.pooling_ratio)
+            elif pool_type == 'SelfAtt':
+                self.pools.append(
+                    SAB(hidden_channels, hidden_channels,
+                        num_heads, conv=None, ln=ln)
+                )
+        self.pools.append(torch.nn.Linear(hidden_channels, out_channels))
+
+    def forward(self, x, batch, edge_index=None):
```
Comment (rusty1s) [inline torch_geometric/nn/glob/gmt.py:173]:
```suggestion
    def forward(self, x: Tensor, batch: Tensor, edge_index: Optional[Tensor]=None) -> Tensor:
```

File: torch_geometric/nn/glob/gmt.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+            self.S.repeat(X.size(0), 1, 1), X, graph, attention_mask
+        )
+
+
+class GraphMultisetTransformer(torch.nn.Module):
+    r"""The global pooling operator, Graph Multiset Transformer,
+    from the `"Accurate Learning of Graph Representations
+    with Graph Multiset Pooling" <https://arxiv.org/abs/2102.11533>`_ paper.
+
+    In short, nodes of the entire graph is clustered with attention-based
```
Comment (rusty1s) [inline torch_geometric/nn/glob/gmt.py:95]:
```suggestion
    In short, nodes of the entire graph are clustered via attention-based
```

**Review by rusty1s — APPROVED** [review]
This looks excellent. I only have minor comments. Thank you very much!

File: torch_geometric/nn/glob/gmt.py (hunk 1 above)
Comment (JinheonBaek) [reply torch_geometric/nn/glob/gmt.py]: Dear @rusty1s
I have added the type hints.

Thanks for the great suggestion!

**Comment by JinheonBaek** [comment]
I have added few more commits to refelct the helpful suggestions from @rusty1s.

**Comment by rusty1s** [comment]
This is excellent. Thank you!

**Comment by JinheonBaek** [comment]
> This is excellent. Thank you!

I have learned a lot from this PR and also your thoughtful clean-up of initial codes.
Thanks! :)

**Pull request merged**